sw/host/opentitanlib/src/bootstrap/primitive.rs

```diff
@@ -0,0 +1,160 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+use anyhow::Result;
+use mundane::hash::{Digest, Hasher, Sha256};
+use std::time::Duration;
+use zerocopy::AsBytes;
+
+use crate::bootstrap::{BootstrapOptions, UpdateProtocol};
+use crate::io::spi::{Target, Transfer};
+
+#[derive(AsBytes, Debug, Default)]
+#[repr(C)]
+struct FrameHeader {
+    hash: [u8; Frame::HASH_LEN],
+    frame_num: u32,
+    flash_offset: u32,
+}
+
+#[derive(AsBytes, Debug)]
+#[repr(C)]
+struct Frame {
+    header: FrameHeader,
+    data: [u8; Frame::DATA_LEN],
+}
+
+impl Default for Frame {
+    fn default() -> Self {
+        Frame {
+            header: Default::default(),
+            data: [0xff; Frame::DATA_LEN],
+        }
+    }
+}
+
+impl Frame {
+    const EOF: u32 = 0x8000_0000;
+    const DATA_LEN: usize = 2048 - std::mem::size_of::<FrameHeader>();
+    const HASH_LEN: usize = 32;
+
+    /// Computes the hash in the header.
+    fn header_hash(&self) -> [u8; Frame::HASH_LEN] {
+        let frame = self.as_bytes();
+        let sha = Sha256::hash(&frame[Frame::HASH_LEN..]);
+        sha.bytes()
+    }
+
+    /// Computes the hash over the entire frame.
+    fn frame_hash(&self) -> [u8; Frame::HASH_LEN] {
+        let sha = Sha256::hash(self.as_bytes());
+        sha.bytes()
+    }
+
+    /// Creates a sequence of frames based on a `payload` binary.
+    fn from_payload(payload: &[u8]) -> Vec<Frame> {
+        let mut frames = Vec::new();
+        let last_frame = (payload.len() + Frame::DATA_LEN - 1) / Frame::DATA_LEN - 1;
+        for (i, chunk) in payload.chunks(Frame::DATA_LEN).enumerate() {
+            let mut frame = Frame::default();
+            frame.header.frame_num = i as u32;
+            if i == last_frame {
+                frame.header.frame_num |= Frame::EOF;
+            }
+            frame.header.flash_offset = (i * Frame::DATA_LEN) as u32;
+            frame.data[..chunk.len()].copy_from_slice(chunk);
+            frame.header.hash = frame.header_hash();
```

Comment:
Do you wanna give writing this as a struct literal (except the hash bit, obvs) a shot?

Reply:
Done.  I'm not really sure that it's any better.

One of the things I'm relying on `default` for is to file the `data` portion of the frame with `0xFF` so that any short last chunk is filled with `0xFF` to the end of frame.  Is there a better way to deal with a possibly short slice while initializing `frame.data`?